torch/nn/modules/adaptive.py

```diff
@@ -0,0 +1,207 @@
+
+    Shape:
+        - Input: :math:`(N, in\_features)`
+        - Target: :math:`(N)` where each value is `0 <= targets[i] <= C - 1`
+        - Output: :math:`(N)`
+
+    .. _Efficient softmax approximation for GPUs:
+        https://arxiv.org/abs/1609.04309
+
+    .. _Zipf's law:
+        https://en.wikipedia.org/wiki/Zipf%27s_law
+    """
+
+    def __init__(self, in_features, n_classes, cutoffs, div_value=2.,
+                 return_logprob=False):
+        super(AdaptiveLogSoftmax, self).__init__()
+
+        cutoffs = list(cutoffs)
+
+        if (cutoffs != sorted(cutoffs)):
+            raise ValueError('Cutoffs should be a list of unique, positive in')
+
+        if (cutoffs != sorted(cutoffs)) \
+                or (max(cutoffs) >= (n_classes - 1)) \
+                or (len(set(cutoffs)) != len(cutoffs)):
+
+            raise ValueError("Cutoffs should be a sequence of unique, positive "
+                             "integers sorted in an increasing order, where "
+                             "each value is between 1 and n_classes-1")
+
+        self.return_logprob = return_logprob
+        self.in_features = in_features
+        self.n_classes = n_classes
+        self.cutoffs = cutoffs + [n_classes]
+        self.div_value = div_value
+
+        self.shortlist_size = self.cutoffs[0]
+        self.n_clusters = len(self.cutoffs) - 1
+        self.head_size = self.shortlist_size + self.n_clusters
+
+        self.head = Linear(self.in_features, self.head_size)
+        self.tail = ModuleList()
+
+        for i in range(self.n_clusters):
+
+            hsz = int(self.in_features // (self.div_value ** (i + 1)))
+            osz = self.cutoffs[i + 1] - self.cutoffs[i]
+
+            projection = Sequential(
+                Linear(self.in_features, hsz, bias=False),
+                Linear(hsz, osz)
+            )
+
+            self.tail.append(projection)
+
+    def reset_parameters(self):
+        self.head.reset_parameters()
+        for i2h, h2o in self.tail:
+            i2h.reset_parameters()
+            h2o.reset_parameters()
+
+    def forward(self, input, target):
+        if input.size(0) != target.size(0):
+            raise RuntimeError('Input and target should have the same size '
+                               'in the batch dimension.')
+
+        used_rows = 0
+        batch_size = target.size(0)
+        out_size = batch_size if (self.return_logprob) else 1
+
+        output = input.new(out_size).zero_()
+        gather_inds = target.new(batch_size).zero_()
+
+        cutoff_values = [0] + self.cutoffs
+        for i in range(len(cutoff_values) - 1):
+
+            low_idx = cutoff_values[i]
+            high_idx = cutoff_values[i + 1]
+
+            target_mask = (target >= low_idx) & (target < high_idx)
+            row_indices = target_mask.nonzero().squeeze()
+
+            if row_indices.numel() == 0:
+                continue
+
+            relative_target = target[target_mask] - cutoff_values[i]
```

Comment:
Can you move this to the `else` block of the `if` below to avoid a kernel launch when `i == 0`?